code/controllers/subsystem/bluespace_exploration.dm

```diff
@@ -0,0 +1,391 @@
+#define CLEAR_TURF_PROCESSING_TIME 600
+
+SUBSYSTEM_DEF(bluespace_exploration)
+	name = "Bluespace Exploration"
+	wait = 1
+	priority = FIRE_PRIORITY_EXPLORATION
+	init_order = INIT_ORDER_BS_EXPLORATION
+
+	var/generating = FALSE
+
+	//Starmap generation
+	var/datum/star_system/current_system = null
+	var/list/star_systems = list()
+	var/list/star_links = list()
+
+	//Ruin generation
+	var/datum/space_level/reserved_bs_level
+	var/list/ruin_templates = list()
+	var/obj/docking_port/stationary/away_mission_port
+
+	//=====Factions=====
+	// factions[datum] = new datum
+	var/list/factions = list()
+
+	//=====Ship Tracking=====
+	var/list/spawnable_ships = list()
+	//All ships that are able to use the random generation must be tracked
+	var/list/tracked_ships = list()
+
+	//=====Ship Processing Queue=====
+	// ! Tracks a list of all the ships requesting transit
+	// ! Key (string) - shuttle_id
+	// ! Value (/datum/data_holder/bluespace_exploration) - generation settings
+	var/list/ship_traffic_queue = list()	//The queue for generation is disguised as 'bluespace traffic control'
+
+	//=====Z-Level Wiping=====
+	//Are we currently processing a z-level wipe
+	var/wiping_z_level = FALSE
+	//The target z-level of the wipe
+	var/target_z_level = 0
+	//A list of all turfs on the z-leves divided into equal parts depending on how many ticks we have to wipe
+	var/list/wiping_divided_turfs
+	//The turf group we are currently wiping
+	var/wipe_process_num  = 0
+	//Data holder for turf wiping
+	var/datum/data_holder/bluespace_exploration/wipe_data_holder
+	//The queue for z-levels to wipe
+	var/list/z_level_queue
+
+/datum/controller/subsystem/bluespace_exploration/Initialize(start_timeofday)
+	z_level_queue = list()
+	. = ..()
+	//Create factions
+	for(var/faction_datum in subtypesof(/datum/faction))
+		factions[faction_datum] = new faction_datum
+
+/datum/controller/subsystem/bluespace_exploration/fire(resumed = 0)
+	if(times_fired % 50 == 0)
+		for(var/ship_key in tracked_ships)
+			var/datum/ship_datum/SD = tracked_ships[ship_key]
+			SD.update_ship()
+			if(QDELETED(SD))
+				tracked_ships -= ship_key
+			CHECK_TICK
+		//Keep doing this just in case
+		initiate_queued_warp()
+		CHECK_TICK
+	if(!wiping_z_level && LAZYLEN(z_level_queue))
+		for(var/z_level_id in z_level_queue)	//There is probably a better way to do this
+			var/z_level = text2num(z_level_id)
+			wipe_z_level(z_level, z_level_queue[z_level_id])
+			break
```

Comment:
```suggestion
		wipe_z_level(z_level, z_level_queue[text2num(z_level_id[1])])
```

Reply:
Holy shit, that original code hurts me

Reply:
Assuming zlevels are stored as text instead of nums for some reason... In which case... why are they stored as text?

Reply:
Eh this code is wrong but you get the idea?

Reply:
wait this code doesn't work, let me go fix it